### DISCUSSION TITLE
✅ Raw request content types

### DISCUSSION THREAD

**Discussion by master-chu** [RobinTail/express-zod-api#869] [Ideas]
### Problem
Currently it looks like the only supported request (input) content types are `application/json` and `multipart/form-data`. My understanding is that if the input schema contains a `z.upload()`, then it will be form data, otherwise JSON.

It would be great if arbitrary types could be specified. Express supports this with their [`raw` middleware](https://expressjs.com/en/api.html#express.raw). I have an endpoint that accepts `application/octet-stream` data as the entire request body, and we were actually able to make this work with `express-api-zod` by intercepting the `response.body` object and forwarding it to the handler function. However, in the generated OpenApi docs, the content type is still specified as `application/json`, which makes it impossible to actually call from the Swagger UI.

### Example
Let's say I have an endpoint called like `POST /my-endpoint/1`

In plain express, the endpoint would look like:
```typescript
const myRouter = Router()
myRouter.post(
    '/my-endpoint/:id',
    raw({ type: 'application/octet-stream' }),
    (req, res, next) => {
        console.log(req.params) // { id: 1 }
        console.log(req.body) // <Buffer 68 65 6c 6c 6f>
        // ...
    },
)
```

And in `express-zod-api` it looks like:
```typescript
const myHandler = myEndpointFactory
    .addExpressMiddleware(
        raw({ type: 'application/octet-stream' }),
    )
    .addMiddleware(createMiddleware({
        input: z.object({}),
        middleware: async ({ request }) => ({
            request,
        }),
    }))
    .build({
        method: 'post',
        input: z.object({
            id: z.number(),
        }),
        output: z.object({}),
        async handler({
            input,
            options: { request },
        }) {
            console.log(input) // { id: 1 }
            console.log(request.body) // <Buffer 68 65 6c 6c 6f>
            // ...
        },
    })
```

This works fine, but as I mentioned the generated spec ends up looking like this:

```json
"/my-endpoint/{id}": {
    "post": {
        "requestBody": {
            "content": {
                "application/json": {
                    "schema": {
                        // omitted for brevity
                    }
                }
            }
        }
    }
}
```
where `application/json` ought to be `application/octet-stream`, which changes the interface Swagger gives to test the endpoint.

### Proposal
I don't have a great solution for how to support this, but one possible approach would be similar to the `z.upload` pattern:

```typescript
.build({
        method: 'post',
        input:  z.raw({ type: 'application/octet-stream' }),
        // ...
    })
```
But since path params and request body are typically merged, I don't really know what the resulting `{ input }` object should look like if you have both. Alternatively, if the intended content type could be inferred from the inclusion of a custom middleware, that would allow them to be separated easily.

Thanks in advance!

**Comment by RobinTail** [comment]
Hello @master-chu ,

thank you for the very detailed problem statement.
It's an interesting topic that I need some time to think on.
I'll get back to you once I get enlightened.

**Comment by RobinTail** [comment]
The current implementation on determining the MIME types of the Endpoint (src/endpoint.ts)

```ts
    this.mimeTypes = {
      input: hasUpload(inputSchema) ? [mimeMultipart] : [mimeJson], // <—— this line
      positive: getMimeTypesFromApiResponse(apiResponse.positive),
      negative: getMimeTypesFromApiResponse(apiResponse.negative),
    };
```

**Comment by RobinTail** [comment]
The problem, however, is that the entire body is the file, not its property, like when having uploads supplied using `multipart/form-data`. There should be a middleware (to be implemented) that acts similarly to `express-fileupload`, that places the binary data into a third level property of `request`.

**Comment by RobinTail** [comment]
@master-chu , consider this:

```ts
const app = express();

/** @link https://expressjs.com/en/4x/api.html#express.raw */
app.use(express.raw());  // default type is application/octet-stream

// custom middleware that places the raw data into body.data
app.use((req, res, next) => {
  if (req.is("application/octet-stream")) {
    req.body = { data: req.body }
  }
  next();
});

// custom express app approach
attachRouting(config, app); // express-zod-api 
```

> Swagger

On that I'll have to think a bit more.

**Reply by RobinTail** [reply]
I also found out that `verify` option of `express.raw()` argument can be utilized to place the data into `req.body.data` prop.

**Comment by RobinTail** [comment]
I think I could make a config option to:
- accept the list of mime types
- enable `express.raw` middleware on them
- activate the custom middleware on placing raw data into `req.body.data` as shown above
- and finally adjust the mime types of the `Endpoint` which will lead to fixed documentation

That's the idea. Should be clean enough.
@master-chu

**Comment by RobinTail** [comment]
Check this out, @master-chu : #1281

**Reply by master-chu** [reply]
@RobinTail This is awesome, looks good to me!

**Comment by RobinTail** [comment]
🚀  The feature released in v14.2.0, @master-chu 
https://ez.robintail.cz/v14.2.0/accepting-raw-data
